### PR TITLE
feat(query): new table engine: RANDOM.

### LINKED ISSUE #5811: Feature: random Table Engine
**Summary**

This is very useful for testing.

```
create table t ( a int, b varchar, ....) Engine = Random;

select * from t limit 10; -- will generate ten rows random data .
```

Thus we can use this engine to generate random data and use them to test other features.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
Signed-off-by: RinChanNOWWW <hzy427@gmail.com>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Implemented a new table engine: RANDOM.

select with limit:

image

select without limit:

image

## Changelog

- New Feature

## Related Issues

Fixes #5811

### CODE REVIEW

File: query/src/storages/random/random_table.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                }
+                match push_downs.limit {
+                    Some(limit) => limit,
+                    None => block_size,
+                }
+            }
+            None => block_size,
+        };
+
+        SyncSourcer::create(ctx, output, RandomSource {
```
Comment (RinChanNOWWW) [inline query/src/storages/random/random_table.rs:138]: Should I implement a multi sourcer? Like:

https://github.com/datafuselabs/databend/blob/652bb413c7e5277928d7c9110ca7198bb9098023/query/src/storages/memory/memory_table.rs#L225-L238

File: query/src/storages/random/random_table.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        loop {
+            for (col, pack) in packs.iter_mut().enumerate() {
+                let col_data_type = self.schema.field(col).data_type();
+                pack.append_data_value(col_data_type.random_value(), &format)?;
+            }
+
+            rows += 1;
+            self.rows += 1;
+
+            // Only generate at most `block_size` rows in one turn.
```
Comment (RinChanNOWWW) [inline query/src/storages/random/random_table.rs:175]: I don't know how to generate rows with exactly `block_size` of physical memory size, so I simply generate `block_size` rows one time. Is this right?

**Comment by RinChanNOWWW** [comment]
Besides the comment above, I have other questions:

- I only implemented `read2`, and how should I write unit test for it?
- I don't implement `random_value` for semi-structured data types, is it OK?

File: common/datavalues/src/types/type_date.rs [hunk 3]
```diff
@@ -68,6 +69,12 @@ impl DataType for DateType {
         DataValue::Int64(0)
     }
 
+    fn random_value(&self) -> DataValue {
```
Comment (sundy-li) [inline common/datavalues/src/types/type_date.rs:72]: Better to have `random_column(length: usize) -> ColumnRef`.

Comment (sundy-li) [reply common/datavalues/src/types/type_date.rs:72]: This can have better performance.

Comment (sundy-li) [reply common/datavalues/src/types/type_date.rs:72]: let it = (0..length).map(...);
let col = I32::from_own_iterator(it);

File: query/src/storages/random/random_table.rs (hunk 1 above)
Comment (sundy-li) [reply query/src/storages/random/random_table.rs:138]: Yes,better to have.

Comment (RinChanNOWWW) [reply query/src/storages/random/random_table.rs:138]: If so, how many rows should one source generate one time? Will the sources generate rows parallelly or concurrently?

Comment (sundy-li) [reply query/src/storages/random/random_table.rs:136]: You can split the total rows into n Parts where n is the num of threads.

refer:  read_partitions  function  in https://github.com/datafuselabs/databend/blob/9261d9cc06d959fdca5516de7cd1381f5c9509df/query/src/table_functions/numbers_table.rs

File: query/src/storages/random/random_table.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        plan: &ReadDataSourcePlan,
+        pipeline: &mut NewPipeline,
+    ) -> Result<()> {
+        let settings = ctx.get_settings();
+        let block_size = settings.get_max_block_size()? as usize;
+        let mut output_schema = self.table_info.schema();
+        let push_downs = plan.push_downs.clone();
+        // If extras.push_downs is None or extras.push_down.limit is None,
+        // set limit to `max_block_size`.
+        let limit = match push_downs {
```
Comment (sundy-li) [inline query/src/storages/random/random_table.rs:91]: This is incorrect, we should add these codes into `read_partitions`, thus we will have `Vec<Partitions>`.

Then we can have a `RandomSource`  like `NumbersSource`

Comment (RinChanNOWWW) [reply query/src/storages/random/random_table.rs:91]: OK

**Review by sundy-li — APPROVED** [review]

**Comment by BohuTANG** [comment]
Good job, thank you! @RinChanNOWWW

**Pull request merged**